### ISSUE TITLE
Build directory name not allowed to have wildcards in them

### ISSUE THREAD

**Issue by pcanal** [root-project/root#13731]
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

If the build directory is named (for example): `master.c++20` we get (zillions of times the lines):
```
CMake Error at cmake/modules/RootMacros.cmake:336 (list):
  list sub-command FILTER, mode REGEX failed to compile regex
  "^/Users/pcanal/root_working/builds/master.c++20/ginclude".
Call Stack (most recent call first):
  cmake/modules/RootMacros.cmake:1361 (ROOT_GENERATE_DICTIONARY)
  core/clingutils/CMakeLists.txt:69 (ROOT_STANDARD_LIBRARY_PACKAGE)

```

We should either fix this or make the error messages clearer (i.e detect the case early and fail clearly)>

### Reproducer

Use a wild card (eg `+`) in the build directory name.

### ROOT version

master.

### Installation method

source

### Operating system

linux

**Comment by bellenot** [comment]
We can escape all special characters, but which ones should we support? Because we might have problems with `${ROOTSYS}` containing those characters. I tried with `master.[c++17](test)` (escaping these characters: `[ ] + . ( )`, but then I got many warnings like:
```
CMake Warning (dev) at C:/root-dev/build/x64/master.c++17(test)/ROOTConfig.cmake:67:
  Syntax Warning in cmake code at column 63

  Argument not separated from preceding token by whitespace.
Call Stack (most recent call first):
  cmake/modules/SearchInstalledSoftware.cmake:18 (_find_package)
  test/CMakeLists.txt:18 (find_package)
```
And also:
```
CMake Warning:
  Value of ROOTSYS contained a newline; truncating
```

**Comment by vepadulano** [comment]
I propose a possible fix in the linked PR, based on existing CMake discussion, practically we escape every character, CMake is happy to accept `\a` as `a` :)

**Comment by bellenot** [comment]
> I propose a possible fix in the linked PR, based on existing CMake discussion, practically we escape every character, CMake is happy to accept `\a` as `a` :)

Right, but accepting any character will not help, see my previous comment

**Comment by vepadulano** [comment]
I don't get those errors with the linked PR. But I get different ones during the build
```
[16/7002] Copying /home/vpadulan/Programs/rootproject/rootsrc/tutorials/graphics/crown.C
FAILED: tutorials/graphics/crown.C /home/vpadulan/Programs/rootproject/rootbuild/master.[c++17](test)/tutorials/graphics/crown.C 
cd "/home/vpadulan/Programs/rootproject/rootbuild/master.[c++17](test)" && /usr/bin/cmake -E copy_if_different /home/vpadulan/Programs/rootproject/rootsrc/tutorials/graphics/crown.C /home/vpadulan/Programs/rootproject/rootbuild/master.[c++17](test)/tutorials/graphics/crown.C
/bin/sh: -c: line 1: syntax error near unexpected token `('
/bin/sh: -c: line 1: `cd "/home/vpadulan/Programs/rootproject/rootbuild/master.[c++17](test)" && /usr/bin/cmake -E copy_if_different /home/vpadulan/Programs/rootproject/rootsrc/tutorials/graphics/crown.C /home/vpadulan/Programs/rootproject/rootbuild/master.[c++17](test)/tutorials/graphics/crown.C'
```

**Comment by bellenot** [comment]
That's my point. We should not allow any special character, or we have to properly handle them through the whole CMake build system, also with the tests and roottest...

**Comment by vepadulano** [comment]
I agree, also doing some more research led me to [this](https://gitlab.kitware.com/cmake/cmake/-/issues/16207).

The CMake maintainers themselves state that the best workaround for "cmake not working properly when build directories include special characters" is "do not use those special characters". So probably we should not do anything about this

**Comment by bellenot** [comment]
Or test what can be used. The `master.c++17` seems OK so far, at least on Windows (it's building right now)

**Comment by guitargeek** [comment]
Maybe also connected to this issue:
https://github.com/root-project/root/issues/7311

**Comment by bellenot** [comment]
BTW, even after fixing `RootMacros.cmake`, there are still a lot of those warnings:
```
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
```
coming from somewhere else (trying to figure out)

**Comment by pcanal** [comment]
> The CMake maintainers themselves state that the best workaround for "cmake not working properly when build directories include special characters" is "do not use those special characters". So probably we should not do anything about this

I think at this point we should explicitly reject if the directory name contains unsupported characters.  Not doing anything is in my opinion not an option as it still result in error message but with 2 deadly flaws: it floods the screen with zillions of repetitions and requires to (after finding the "right" location) the line of CMake to figure out why it complains.

**Comment by arabusov** [comment]
@pcanal it would be nice if the cmake invocation stopped immediately if the path contain "wrong" characters.

**Comment by ferdymercury** [comment]
Related discussion: https://stackoverflow.com/questions/13569660/cmake-escaping-symbols-inside-a-variable-in-regular-expressions
